specs/Gem.spec

```diff
@@ -0,0 +1,206 @@
+    mathint total_supply_after = totalSupply();
+
+    assert balance_recip_after == balance_recip_before + amount, "recip balance did not increase by mint amount";
+    assert total_supply_after == total_supply_before + amount, "total supply did not increase by mint amount";
+
+}
+
+rule mintMustNeverOverflow {
+        address recip; uint amount;
+    
+    env e;
+    mathint total_supply = totalSupply();
+    require(total_supply + amount > max_uint256);
+
+    mint@withrevert(e, recip, amount);
+    assert lastReverted, "mint must revert if total supply overflows";
+}
+
+rule burnMustAlwaysDecreaseBalanceAndTotalSupply() {
+        address burn; uint amount;
+    env e;
+    require(balanceOf(burn) <= amount);
+    require(balanceOf(burn) <= totalSupply());
+
+    mathint total_supply_before = totalSupply();
+    mathint balance_before = balanceOf(burn);
+    burn(e, burn, amount);
+
+    mathint total_supply_after = totalSupply();
+    mathint balance_after = balanceOf(burn);
+
+    assert total_supply_after == total_supply_before - amount, "totalSupply did not decrease by burn amount";
+    assert balance_after == balance_before - amount, "usr balance did not decrease by burn amount";
+}
+
+rule onlyMintAndBurnCanChangeTotalSupply(address target, uint amt) {
+
+    env e; method f; calldataarg args;
+    mathint total_supply_before = totalSupply();
+
+    f@withrevert(e, args);
+
+    mathint total_supply_after = totalSupply();
+
+    assert total_supply_before != total_supply_after 
+        => 
+    (f.selector == mint(address,uint).selector || f.selector == burn(address,uint).selector),
+    "function other than mint or burn changed total supply!";
+
+
+}
+
+rule burnMustNeverUnderflow {
+        address recip; uint amount;
+    
+    env e;
+    require(balanceOf(recip) < amount);
+
+    burn@withrevert(e, recip, amount);
+    assert lastReverted, "burn must revert if total supply underflows";
+}
+
+rule mintAndBurnMustAlwaysRequireWard {
+        address other; uint amount;
+
+    env e;
+    address sender = e.msg.sender;
+    ward(e, sender, false);
+
+    mint@withrevert(e, other, amount);
+    assert lastReverted, "mint did not revert with non-ward sender";
+
+    burn@withrevert(e, other, amount);
+    assert lastReverted, "burn did not revert with non-ward sender";
+}
+
+rule wardSpec {
+        address other_ward;
+    
+    env e;
+    address sender = e.msg.sender;
+    require(wards(sender) == true); // for this spec, assume sender is already a ward
+
+    ward(e, other_ward, true);
+    assert wards(other_ward) == true;
+
+    ward(e, other_ward, false);
+    assert wards(other_ward) == false;
+
+}
+
+rule approveSpec {
+        address spender; uint amount;
+    
+    env e;
+
+    approve(e, spender, amount);
+
+    assert allowance(e.msg.sender, spender) == amount, "spender allowance does not match intended amount";
+}
+
+rule transferFromSpec(address src, address dst, uint amt) {
+    require(src != dst);
```

Comment:
don't completely ignore this case pls, it's a common bug that's actually mentioned in the prover examples